### PR TITLE
[dynamo] lift the constraint that cannot make_fx a dynamo compiled function

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109734
* #109916

Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Comment by ydwu4** [comment]
The current implementation didn't handle expressions that are generated from input symbool correctly: to do that, we'll need to propagate the expressions on both outer shape_env and inner shape_env. It's requires a lot of effort.

The next plan is run the generated shape_env guard once with input symbol (from outer shape_env) to install the guards in outer shape_env

**Comment by ezyang** [comment]
do you want review now or later

**Comment by ydwu4** [comment]
> do you want review now or later

Later today! I got some ideas. Will ping you when it's ready

File: torch/_dynamo/eval_frame.py [hunk 1]
```diff
@@ -377,9 +378,14 @@ def get_compiler_config():
...
         @functools.wraps(fn)
         def _fn(*args, **kwargs):
+            any_arg_is_proxy = any(
+                isinstance(arg, torch.fx.Proxy)
+                for arg in itertools.chain(args, kwargs.values())
+            )
             if (
                 not isinstance(self, DisableContext)
                 and torch.fx._symbolic_trace.is_fx_tracing()
+                and any_arg_is_proxy
```
Comment (ezyang) [inline torch/_dynamo/eval_frame.py:388]: What's going on here?

**Comment by ezyang** [comment]
I am not so hot on just removing this assert. Maybe we can treat "eager" dynamo compiled regions differently from conventional (compiled) regions?

File: torch/_dynamo/eval_frame.py (hunk 1 above)
Comment (ydwu4) [reply torch/_dynamo/eval_frame.py:388]: is_fx_tracing flag is also set for make_fx. So we try to detect whether we're in symbolic_trace mode by checking if any arguments are wrapped as proxy.

**Comment by ydwu4** [comment]
> I am not so hot on just removing this assert. Maybe we can treat "eager" dynamo compiled regions differently from conventional (compiled) regions?

Yeah, sounds good. Let me fix it by adding a backend check as well.

File: test/dynamo/test_misc.py [hunk 2]
```diff
@@ -4462,6 +4465,66 @@ def f(x):
...
+                    """\
+def forward(self, x_1):
+    sin = torch.ops.aten.sin.default(x_1);  x_1 = None
+    return sin""",
+                )
+            self.assertExpectedInline(_get_guard_exprs(gms[0]), """Eq(s0, 3)""")
+            for gm in gms[1:]:
+                self.assertExpectedInline(
+                    _get_guard_exprs(gm),
+                    """Eq(Piecewise((1, Eq(s0, 3)), (0, True)), 1)""",
```
Comment (ezyang) [inline test/dynamo/test_misc.py:4510]: The RHS of the Piecewise is constant, why isn't this simplified 🤔

**Review by ezyang — CHANGES_REQUESTED** [review]
Please more carefully review your expect test output, it still looks wrong

**Pull request closed without merging**